src/include/migraphx/op/loop.hpp

```diff
@@ -0,0 +1,139 @@
+#include <migraphx/argument.hpp>
+#include <migraphx/functional.hpp>
+#include <migraphx/config.hpp>
+#include <migraphx/module.hpp>
+#include <migraphx/run_loop.hpp>
+#include <migraphx/ranges.hpp>
+#include <cmath>
+#include <string>
+#include <utility>
+#include <set>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+namespace op {
+
+struct loop
+{
+    int64_t max_iter_num = 0;
+
+    template <class Self, class F>
+    static auto reflect(Self& self, F f)
+    {
+        return pack(f(self.max_iter_num, "max_iter_num"));
+    }
+
+    std::string name() const { return "loop"; }
+
+    shape compute_shape(const std::vector<shape>& inputs, std::vector<module_ref> mods) const
+    {
+        check_shapes{inputs, *this}.standard();
+        if(mods.size() != 1)
+        {
+            MIGRAPHX_THROW("LOOP: operator should have one submodule.");
+        }
+
+        const auto& mod     = mods.front();
+        auto mod_out_shapes = mod->get_output_shapes();
+        auto dep_param_num  = inputs.size() - 2;
+        // first two names -- iter_num and cond_var -- are not counted
+        mod_out_shapes.erase(mod_out_shapes.begin());
+        std::vector<shape> ins_out_shapes(mod_out_shapes.begin(),
+                                          mod_out_shapes.begin() + dep_param_num);
+        mod_out_shapes.erase(mod_out_shapes.begin(), mod_out_shapes.begin() + dep_param_num);
+        for(const auto& out_s : mod_out_shapes)
+        {
+            auto lens = out_s.lens();
+            lens.insert(lens.begin(), max_iter_num);
+            ins_out_shapes.push_back({out_s.type(), lens});
+        }
+
+        return shape(ins_out_shapes);
+    }
+
+    struct ref_loop
+    {
+        int64_t max_iter_num = 0;
+
+        template <class T>
+        void copy_arg(context&, const argument& arg, T& var, bool from_var) const
+        {
+            char* ptr = reinterpret_cast<char*>(&var);
+            if(from_var)
+            {
+                std::copy(ptr, ptr + arg.get_shape().bytes(), arg.data());
+            }
+            else
+            {
+                std::copy(arg.data(), arg.data() + arg.get_shape().bytes(), ptr);
+            }
+        }
+
+        void concat_scan_outputs(const std::vector<argument>& mod_scan_outs,
+                                 const std::vector<argument>& scan_outputs,
+                                 const int iter) const
+        {
+            assert(mod_scan_outs.size() == scan_outputs.size());
+            for(auto i : range(mod_scan_outs.size()))
+            {
+                const auto& mod_out  = mod_scan_outs.at(i);
+                const auto& scan_out = scan_outputs.at(i);
+
+                auto* in_data        = mod_out.data();
+                auto* out_data       = scan_out.data();
+                std::size_t out_size = mod_out.get_shape().bytes();
+                std::copy(in_data, in_data + out_size, out_data + iter * out_size);
+            }
+        }
```

Comment:
This looks more like it appends the `mod_scan_outs`, Maybe we should rename this function `append`?